### PR TITLE
Index service rest interface

### PR DESCRIPTION
**Pull request by jjelschen**
Simple JSON (de-)serializer for arbitrary IEObjectDescription elements.

### CODE REVIEW

File: src/main/java/org/testeditor/web/xtext/index/XtextIndexHelloWorldApplication.xtend [hunk 1]
```diff
@@ -30,7 +34,15 @@ class XtextIndexHelloWorldApplication extends Application<XtextIndexHelloWorldCo
 	}
 
 	override initialize(Bootstrap<XtextIndexHelloWorldConfiguration> bootstrap) {
-		// nothing to do yet
+		registerCustomEObjectSerializer(bootstrap)
+	}
+
+	def registerCustomEObjectSerializer(Bootstrap<XtextIndexHelloWorldConfiguration> bootstrap) {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/web/xtext/index/XtextIndexHelloWorldApplication.xtend]: can be private, right?

File: src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionDeserializer.xtend [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import org.eclipse.xtext.resource.persistence.SerializableEObjectDescription
+
+import static org.testeditor.web.xtext.index.serialization.EObjectDescriptionSerialization.EOBJECT_URI__FIELD_NAME
+import static org.testeditor.web.xtext.index.serialization.EObjectDescriptionSerialization.QUALIFIED_NAME__DELIMITER
+import static org.testeditor.web.xtext.index.serialization.EObjectDescriptionSerialization.QUALIFIED_NAME__FIELD_NAME
+import static org.testeditor.web.xtext.index.serialization.EObjectDescriptionSerialization.URI__FIELD_NAME
+
+class EObjectDescriptionDeserializer extends StdDeserializer<IEObjectDescription> {
+
+	new(Class<?> vc) {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionDeserializer.xtend]: Is this ctor necessary?

File: src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionDeserializer.xtend [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+		result.qualifiedName = QualifiedName.create(
+			node.getTextValue(QUALIFIED_NAME__FIELD_NAME).split(Pattern.quote(QUALIFIED_NAME__DELIMITER)))
+
+		return result
+	}
+
+	private def eClassFromURIString(String uriString) {
+		val uri = URI.createURI(uriString)
+		val ePackage = EPackage.Registry.INSTANCE.getEPackage(uri.trimFragment.toString)
+		ePackage.eResource.getEObject(uri.fragment) as EClass
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionDeserializer.xtend]: please add explicit return statement (settled convention)

File: src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionDeserializer.xtend [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	}
+
+	private def getTextValue(TreeNode node, String fieldName) {
+		(node.get(fieldName) as TextNode).textValue
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionDeserializer.xtend]: please add explicit return statement (settled convention)

File: src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionDeserializer.xtend (hunk 2 above)
Comment (gunther-bachmann) [reply src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionDeserializer.xtend]: Why not define ctor without parameters, passing super IEObjectDescription?

File: src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionSerializer.xtend [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import org.eclipse.xtext.resource.IEObjectDescription
+
+import static org.testeditor.web.xtext.index.serialization.EObjectDescriptionSerialization.EOBJECT_URI__FIELD_NAME
+import static org.testeditor.web.xtext.index.serialization.EObjectDescriptionSerialization.QUALIFIED_NAME__DELIMITER
+import static org.testeditor.web.xtext.index.serialization.EObjectDescriptionSerialization.QUALIFIED_NAME__FIELD_NAME
+import static org.testeditor.web.xtext.index.serialization.EObjectDescriptionSerialization.URI__FIELD_NAME
+
+class EObjectDescriptionSerializer extends StdSerializer<IEObjectDescription> {
+
+	new(Class<IEObjectDescription> t) {
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionSerializer.xtend]: (see ctor comment for deserializer)

File: src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionSerializer.xtend [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+		super(t)
+	}
+
+	override serialize(IEObjectDescription value, JsonGenerator gen,
```
Comment (gunther-bachmann) [inline src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionSerializer.xtend]: rename value to eObjectDescription

File: src/test/java/org/testeditor/web/xtext/index/api/EObjectDescriptionSerializationTest.xtend [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+	def static serializersHaveToBeRegistered() {
+		val customSerializerModule = new SimpleModule
+		customSerializerModule.addSerializer(IEObjectDescription, new EObjectDescriptionSerializer(IEObjectDescription))
+		customSerializerModule.addDeserializer(IEObjectDescription,
+			new EObjectDescriptionDeserializer(IEObjectDescription))
+		mapper.registerModule(customSerializerModule)
+	}
+
+	def static metamodelMustBeKnown() {
```
Comment (gunther-bachmann) [inline src/test/java/org/testeditor/web/xtext/index/api/EObjectDescriptionSerializationTest.xtend]: naming is misleading, how about something like 'makeXtextModelKnown'

File: src/test/java/org/testeditor/web/xtext/index/api/EObjectDescriptionSerializationTest.xtend [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+class EObjectDescriptionSerializationTest {
+	static extension ObjectMapper mapper = Jackson.newObjectMapper
+
+	@BeforeClass
+	static def void ensureTestAssumptions() {
+		serializersHaveToBeRegistered()
+		metamodelMustBeKnown()
+	}
+
+	def static serializersHaveToBeRegistered() {
```
Comment (gunther-bachmann) [inline src/test/java/org/testeditor/web/xtext/index/api/EObjectDescriptionSerializationTest.xtend]: name is misleading, how about registerSerializers

File: src/test/java/org/testeditor/web/xtext/index/api/EObjectDescriptionSerializationTest.xtend [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+		val actual = description.writeValueAsString
+
+		// then
+		assertThat(actual).isEqualTo(expected)
+	}
+
+	@Test
+	def void shouldDeserializeFromJSON() throws Exception {
+		// given
+		val json = "fixtures/eObjectDescription.json".fixture
```
Comment (gunther-bachmann) [inline src/test/java/org/testeditor/web/xtext/index/api/EObjectDescriptionSerializationTest.xtend]: sampleEObjectDescription would be more to the point, what do you think?

File: src/main/java/org/testeditor/web/xtext/index/XtextIndexHelloWorldApplication.xtend (hunk 1 above)
Comment (jjelschen) [reply src/main/java/org/testeditor/web/xtext/index/XtextIndexHelloWorldApplication.xtend]: yupp

File: src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionDeserializer.xtend (hunk 2 above)
Comment (jjelschen) [reply src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionDeserializer.xtend]: I need a public constructor, but you're right, the argument is completely redundant.

File: src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionSerializer.xtend (hunk 6 above)
Comment (jjelschen) [reply src/main/java/org/testeditor/web/xtext/index/serialization/EObjectDescriptionSerializer.xtend]: ... also renamed "gen" to "generator" in the process.

File: src/test/java/org/testeditor/web/xtext/index/api/EObjectDescriptionSerializationTest.xtend (hunk 8 above)
Comment (jjelschen) [reply src/test/java/org/testeditor/web/xtext/index/api/EObjectDescriptionSerializationTest.xtend]: Well, its about the *assumptions* the test has to make about the environment it is run in (global preconditions / "givens"). So if you read the method `ensureTestAssumptions`, it documents these assumptions pretty well (or so I thought). Anyways, your suggestions probably work just as well :)

File: src/test/java/org/testeditor/web/xtext/index/api/EObjectDescriptionSerializationTest.xtend (hunk 9 above)
Comment (jjelschen) [reply src/test/java/org/testeditor/web/xtext/index/api/EObjectDescriptionSerializationTest.xtend]: ok, but the test says "should deserialize from *JSON*"...

**Pull request merged**